From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Zach Brown <zach.brown@destroystokyo.com>
Date: Tue, 1 Mar 2016 13:09:16 -0600
Subject: [PATCH] Configurable baby zombie movement speed


diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index 098bd3fba867c0e4c6c58748aa6e2e632737a948..912611cf1aeccf5a82a789aab07d76723d4357cc 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -73,4 +73,15 @@ public class PaperWorldConfig {
         log("Max height for cactus growth " + cactusMaxHeight + ". Max height for reed growth " + reedMaxHeight);
 
     }
+
+    public double babyZombieMovementModifier;
+    private void babyZombieMovementModifier() {
+        babyZombieMovementModifier = getDouble("baby-zombie-movement-modifier", 0.5D);
+        if (PaperConfig.version < 20) {
+            babyZombieMovementModifier = getDouble("baby-zombie-movement-speed", 0.5D);
+            set("baby-zombie-movement-modifier", babyZombieMovementModifier);
+        }
+
+        log("Baby zombies will move at the speed of " + babyZombieMovementModifier);
+    }
 }
diff --git a/src/main/java/net/minecraft/entity/monster/ZombieEntity.java b/src/main/java/net/minecraft/entity/monster/ZombieEntity.java
index 9a2293564364d0774bca25806ba2858fe7c11d6e..daef0f04d5e00fff5bafb7ae70c67521502aed88 100644
--- a/src/main/java/net/minecraft/entity/monster/ZombieEntity.java
+++ b/src/main/java/net/minecraft/entity/monster/ZombieEntity.java
@@ -79,7 +79,7 @@ import org.bukkit.event.entity.EntityTransformEvent;
 public class ZombieEntity extends MonsterEntity {
 
     private static final UUID field_110187_bq = UUID.fromString("B9766B59-9566-4402-BC1F-2EE2A276D836");
-    private static final AttributeModifier field_110188_br = new AttributeModifier(ZombieEntity.field_110187_bq, "Baby speed boost", 0.5D, AttributeModifier.Operation.MULTIPLY_BASE);
+    private final AttributeModifier field_110188_br = new AttributeModifier(ZombieEntity.field_110187_bq, "Baby speed boost", 0.5D, AttributeModifier.Operation.MULTIPLY_BASE); private final AttributeModifier babyModifier = this.field_110188_br; // Paper - remove static - Make baby speed configurable
     private static final DataParameter<Boolean> field_184737_bv = EntityDataManager.func_187226_a(ZombieEntity.class, DataSerializers.field_187198_h);
     private static final DataParameter<Integer> field_184738_bw = EntityDataManager.func_187226_a(ZombieEntity.class, DataSerializers.field_187192_b);
     public static final DataParameter<Boolean> field_204709_bA = EntityDataManager.func_187226_a(ZombieEntity.class, DataSerializers.field_187198_h);
@@ -181,9 +181,9 @@ public class ZombieEntity extends MonsterEntity {
         if (this.field_70170_p != null && !this.field_70170_p.field_72995_K) {
             ModifiableAttributeInstance attributemodifiable = this.func_110148_a(Attributes.field_233821_d_);
 
-            attributemodifiable.func_111124_b(ZombieEntity.field_110188_br);
+            attributemodifiable.func_111124_b(this.babyModifier); // Paper
             if (p_82227_1_) {
-                attributemodifiable.func_233767_b_(ZombieEntity.field_110188_br);
+                attributemodifiable.func_233767_b_(this.babyModifier); // Paper
             }
         }
 
