From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: DoctorDark <doctordark11@gmail.com>
Date: Wed, 16 Mar 2016 02:21:39 -0500
Subject: [PATCH] Configurable end credits


diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index a005c4398336dba3b5e3ef716f44f2eda7306442..88f5a415770da5dfcf7af95045f547e0761adc8b 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -145,4 +145,10 @@ public class PaperWorldConfig {
             }
         }
     }
+
+    public boolean disableEndCredits;
+    private void disableEndCredits() {
+        disableEndCredits = getBoolean("game-mechanics.disable-end-credits", false);
+        log("End credits disabled: " + disableEndCredits);
+    }
 }
diff --git a/src/main/java/net/minecraft/entity/player/ServerPlayerEntity.java b/src/main/java/net/minecraft/entity/player/ServerPlayerEntity.java
index e94acf808188939992fbc687769cff412577cde7..fce56b6cbc82e5669cbeb3c937680de5288f6120 100644
--- a/src/main/java/net/minecraft/entity/player/ServerPlayerEntity.java
+++ b/src/main/java/net/minecraft/entity/player/ServerPlayerEntity.java
@@ -186,7 +186,7 @@ public class ServerPlayerEntity extends PlayerEntity implements IContainerListen
     private long field_143005_bX = Util.func_211177_b();
     private Entity field_175401_bS;
     public boolean field_184851_cj;
-    private boolean field_192040_cp;
+    private boolean field_192040_cp; private void setHasSeenCredits(boolean has) { this.field_192040_cp = has; } // Paper - OBFHELPER
     private final ServerRecipeBook field_192041_cq = new ServerRecipeBook();
     private Vector3d field_193107_ct;
     private int field_193108_cu;
@@ -895,6 +895,7 @@ public class ServerPlayerEntity extends PlayerEntity implements IContainerListen
             this.func_213319_R();
             this.func_71121_q().func_217434_e(this);
             if (!this.field_71136_j) {
+                if (field_70170_p.paperConfig.disableEndCredits) this.setHasSeenCredits(true); // Paper - Toggle to always disable end credits
                 this.field_71136_j = true;
                 this.field_71135_a.func_147359_a(new SChangeGameStatePacket(SChangeGameStatePacket.field_241768_e_, this.field_192040_cp ? 0.0F : 1.0F));
                 this.field_192040_cp = true;
