From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: vemacs <d@nkmem.es>
Date: Thu, 3 Mar 2016 01:19:22 -0600
Subject: [PATCH] All chunks are slime spawn chunks toggle


diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index 68132778ebafbb65077b82df8fcd8cdda319ffab..e2cb804fd40e236337c37d89dfc85a02486c1bf4 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -187,4 +187,9 @@ public class PaperWorldConfig {
     private void disableChestCatDetection() {
         disableChestCatDetection = getBoolean("game-mechanics.disable-chest-cat-detection", false);
     }
+
+    public boolean allChunksAreSlimeChunks;
+    private void allChunksAreSlimeChunks() {
+        allChunksAreSlimeChunks = getBoolean("all-chunks-are-slime-chunks", false);
+    }
 }
diff --git a/src/main/java/net/minecraft/entity/monster/SlimeEntity.java b/src/main/java/net/minecraft/entity/monster/SlimeEntity.java
index b70e8298d537c838a7b3836c6c1b36b173a6060c..b9bb0f4972ae50ce885c5ebdacd75bb772c17be7 100644
--- a/src/main/java/net/minecraft/entity/monster/SlimeEntity.java
+++ b/src/main/java/net/minecraft/entity/monster/SlimeEntity.java
@@ -324,7 +324,7 @@ public class SlimeEntity extends MobEntity implements IMob {
             }
 
             ChunkPos chunkcoordintpair = new ChunkPos(p_223366_3_);
-            boolean flag = SharedSeedRandom.func_205190_a(chunkcoordintpair.field_77276_a, chunkcoordintpair.field_77275_b, ((ISeedReader) p_223366_1_).func_72905_C(), p_223366_1_.getMinecraftWorld().spigotConfig.slimeSeed).nextInt(10) == 0; // Spigot
+            boolean flag = p_223366_1_.getMinecraftWorld().paperConfig.allChunksAreSlimeChunks || SharedSeedRandom.func_205190_a(chunkcoordintpair.field_77276_a, chunkcoordintpair.field_77275_b, ((ISeedReader) p_223366_1_).func_72905_C(), p_223366_1_.getMinecraftWorld().spigotConfig.slimeSeed).nextInt(10) == 0; // Spigot // Paper
 
             if (p_223366_4_.nextInt(10) == 0 && flag && p_223366_3_.func_177956_o() < 40) {
                 return func_223315_a(p_223366_0_, p_223366_1_, p_223366_2_, p_223366_3_, p_223366_4_);
