From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joseph Hirschfeld <joe@ibj.io>
Date: Thu, 3 Mar 2016 02:39:54 -0600
Subject: [PATCH] Change implementation of (tile)entity removal list

use sets for faster removal

diff --git a/src/main/java/net/minecraft/world/World.java b/src/main/java/net/minecraft/world/World.java
index 6e08a1f0608fe45c8adf194b587b4b0e58e59918..47d451ac90b5b32da2c0d1d7e2fedd8bffdc1422 100644
--- a/src/main/java/net/minecraft/world/World.java
+++ b/src/main/java/net/minecraft/world/World.java
@@ -90,7 +90,7 @@ public abstract class World implements IWorld, AutoCloseable {
     public final List<TileEntity> field_147482_g = Lists.newArrayList();
     public final List<TileEntity> field_175730_i = Lists.newArrayList();
     protected final List<TileEntity> field_147484_a = Lists.newArrayList();
-    protected final List<TileEntity> field_147483_b = Lists.newArrayList();
+    protected final java.util.Set<TileEntity> field_147483_b = com.google.common.collect.Sets.newHashSet();
     public final Thread field_217407_c;
     private final boolean field_234916_c_;
     private int field_73008_k;
