From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 28 Apr 2016 00:57:27 -0400
Subject: [PATCH] remove null possibility for getServer singleton

to stop IDE complaining about potential NPE

diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 428278ece23a2c0392e879cec82adb96204d1365..d56f5d9beb41baa21b3e2667c212c1ebfbc7df6c 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -179,6 +179,7 @@ import org.spigotmc.SlackActivityAccountant; // Spigot
 
 public abstract class MinecraftServer extends RecursiveEventLoop<TickDelayedTask> implements ISnooperInfo, ICommandSource, AutoCloseable {
 
+    private static MinecraftServer SERVER; // Paper
     public static final Logger field_147145_h = LogManager.getLogger();
     public static final File field_152367_a = new File("usercache.json");
     public static final WorldSettings field_213219_c = new WorldSettings("Demo World", GameType.SURVIVAL, false, Difficulty.NORMAL, false, new GameRules(), DatapackCodec.field_234880_a_);
@@ -285,6 +286,7 @@ public abstract class MinecraftServer extends RecursiveEventLoop<TickDelayedTask
 
     public MinecraftServer(OptionSet options, DatapackCodec datapackconfiguration, Thread thread, DynamicRegistries.Impl iregistrycustom_dimension, SaveFormat.LevelSave convertable_conversionsession, IServerConfiguration savedata, ResourcePackList resourcepackrepository, Proxy proxy, DataFixer datafixer, DataPackRegistries datapackresources, MinecraftSessionService minecraftsessionservice, GameProfileRepository gameprofilerepository, PlayerProfileCache usercache, IChunkStatusListenerFactory worldloadlistenerfactory) {
         super("Server");
+        SERVER = this; // Paper - better singleton
         this.field_240769_m_ = new TimeTracker(Util.field_211180_a, this::func_71259_af);
         this.field_71304_b = EmptyProfiler.field_219906_a;
         this.field_147147_p = new ServerStatusResponse();
@@ -2152,7 +2154,7 @@ public abstract class MinecraftServer extends RecursiveEventLoop<TickDelayedTask
 
     @Deprecated
     public static MinecraftServer getServer() {
-        return (Bukkit.getServer() instanceof CraftServer) ? ((CraftServer) Bukkit.getServer()).getServer() : null;
+        return SERVER; // Paper
     }
     // CraftBukkit end
 
