From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Tue, 27 Dec 2016 15:02:42 -0500
Subject: [PATCH] String based Action Bar API


diff --git a/src/main/java/net/minecraft/util/Util.java b/src/main/java/net/minecraft/util/Util.java
index cc391e7130f6e574bd66931abe5f21856e145036..669ed1082b1ff1f66403bd261d2e4f730858c210 100644
--- a/src/main/java/net/minecraft/util/Util.java
+++ b/src/main/java/net/minecraft/util/Util.java
@@ -53,7 +53,7 @@ public class Util {
     private static final ExecutorService field_215091_c = func_240979_a_("Main");
     private static final ExecutorService field_240975_f_ = func_240995_n_();
     public static LongSupplier field_211180_a = System::nanoTime;
-    public static final UUID field_240973_b_ = new UUID(0L, 0L);
+    public static final UUID field_240973_b_ = new UUID(0L, 0L); public static final UUID getNullUUID() {return field_240973_b_;} // Paper OBFHELPER
     private static final Logger field_195650_a = LogManager.getLogger();
 
     public static <K, V> Collector<Entry<? extends K, ? extends V>, ?, Map<K, V>> func_199749_a() {
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index eb4ebdccb391b29fd16a6bac8c2c7b6a3fd8faa3..8847e57e7ffbeeeaf17e60016dbcac3234c09aab 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -220,6 +220,24 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     }
 
     // Paper start
+    @Override
+    public void sendActionBar(BaseComponent[] message) {
+        if (getHandle().field_71135_a == null) return;
+        getHandle().field_71135_a.func_147359_a(new STitlePacket(STitlePacket.Type.ACTIONBAR, message, -1, -1, -1));
+    }
+
+    @Override
+    public void sendActionBar(String message) {
+        if (getHandle().field_71135_a == null || message == null || message.isEmpty()) return;
+        getHandle().field_71135_a.func_147359_a(new STitlePacket(STitlePacket.Type.ACTIONBAR, CraftChatMessage.fromStringOrNull(message)));
+    }
+
+    @Override
+    public void sendActionBar(char alternateChar, String message) {
+        if (message == null || message.isEmpty()) return;
+        sendActionBar(org.bukkit.ChatColor.translateAlternateColorCodes(alternateChar, message));
+    }
+
     @Override
     public void setPlayerListHeaderFooter(BaseComponent[] header, BaseComponent[] footer) {
          if (header != null) {
