From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Zach Brown <zach.brown@destroystokyo.com>
Date: Sun, 2 Jul 2017 21:35:56 -0500
Subject: [PATCH] Block player logins during server shutdown


diff --git a/src/main/java/net/minecraft/network/login/ServerLoginNetHandler.java b/src/main/java/net/minecraft/network/login/ServerLoginNetHandler.java
index 3cd1ba47a9a948291af69570a1c82e908a3118f1..18bc079b7f10f2bb0e4e030c8ea72568f9f751e9 100644
--- a/src/main/java/net/minecraft/network/login/ServerLoginNetHandler.java
+++ b/src/main/java/net/minecraft/network/login/ServerLoginNetHandler.java
@@ -65,6 +65,12 @@ public class ServerLoginNetHandler implements IServerLoginNetHandler {
     }
 
     public void func_73660_a() {
+        // Paper start - Do not allow logins while the server is shutting down
+        if (!MinecraftServer.getServer().func_71278_l()) {
+            this.func_194026_b(org.bukkit.craftbukkit.util.CraftChatMessage.fromString(org.spigotmc.SpigotConfig.restartMessage)[0]);
+            return;
+        }
+        // Paper end
         if (this.field_147328_g == ServerLoginNetHandler.State.READY_TO_ACCEPT) {
             this.func_147326_c();
         } else if (this.field_147328_g == ServerLoginNetHandler.State.DELAY_ACCEPT) {
