From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Spottedleaf <Spottedleaf@users.noreply.github.com>
Date: Sun, 2 Sep 2018 19:34:33 -0700
Subject: [PATCH] Make CraftWorld#loadChunk(int, int, false) load unconverted
 chunks


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index 1281e0cad24a39ee463135648f802dda514271b2..f2545b295abb594efa5090a08a746f4b11e51e6c 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -501,7 +501,7 @@ public class CraftWorld implements World {
     @Override
     public boolean loadChunk(int x, int z, boolean generate) {
         org.spigotmc.AsyncCatcher.catchOp("chunk load"); // Spigot
-        IChunk chunk = world.func_72863_F().func_212849_a_(x, z, generate ? ChunkStatus.field_222617_m : ChunkStatus.field_223226_a_, true);
+        IChunk chunk = world.func_72863_F().func_212849_a_(x, z, generate || isChunkGenerated(x, z) ? ChunkStatus.field_222617_m : ChunkStatus.field_223226_a_, true); // Paper
 
         // If generate = false, but the chunk already exists, we will get this back.
         if (chunk instanceof ChunkPrimerWrapper) {
