From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: BillyGalbreath <Blake.Galbreath@GMail.com>
Date: Tue, 4 Sep 2018 15:02:00 -0500
Subject: [PATCH] Expose attack cooldown methods for Player


diff --git a/src/main/java/net/minecraft/entity/player/PlayerEntity.java b/src/main/java/net/minecraft/entity/player/PlayerEntity.java
index 626c5b0e8c2c889abeffdb624467447deb651b15..78d2c4bed7617710631b21cfa127caad19a7d367 100644
--- a/src/main/java/net/minecraft/entity/player/PlayerEntity.java
+++ b/src/main/java/net/minecraft/entity/player/PlayerEntity.java
@@ -2099,6 +2099,7 @@ public abstract class PlayerEntity extends LivingEntity {
         this.field_70180_af.func_187227_b(PlayerEntity.field_192033_bu, p_192031_1_);
     }
 
+    public float getCooldownPeriod() { return this.func_184818_cX(); } // Paper - OBFHELPER
     public float func_184818_cX() {
         return (float) (1.0D / this.func_233637_b_(Attributes.field_233825_h_) * 20.0D);
     }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 7f71ede003b11a6b14963464f288cb9f1c7830dd..af5a1f4f867f1d8186d55ddac1c1ed3b96a7b214 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -1992,6 +1992,20 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         getInventory().setItemInMainHand(hand);
     }
 
+    // Paper start
+    public float getCooldownPeriod() {
+        return getHandle().getCooldownPeriod();
+    }
+
+    public float getCooledAttackStrength(float adjustTicks) {
+        return getHandle().func_184825_o(adjustTicks);
+    }
+
+    public void resetCooldown() {
+        getHandle().func_184821_cY();
+    }
+    // Paper end
+
     // Spigot start
     private final Player.Spigot spigot = new Player.Spigot()
     {
