From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: BillyGalbreath <Blake.Galbreath@GMail.com>
Date: Sat, 6 Oct 2018 21:47:44 -0500
Subject: [PATCH] Allow setting the vex's summoner


diff --git a/src/main/java/net/minecraft/entity/monster/VexEntity.java b/src/main/java/net/minecraft/entity/monster/VexEntity.java
index 1a1891733e1030603bd4a4e04c3a3facfb074875..fcde016d24a53f9d9e6c59b766fe238c25c9fc82 100644
--- a/src/main/java/net/minecraft/entity/monster/VexEntity.java
+++ b/src/main/java/net/minecraft/entity/monster/VexEntity.java
@@ -165,6 +165,7 @@ public class VexEntity extends MonsterEntity {
         this.func_190660_a(1, p_190648_1_);
     }
 
+    public void setOwner(MobEntity entityinsentient) { func_190658_a(entityinsentient); } // Paper - OBFHELPER
     public void func_190658_a(MobEntity p_190658_1_) {
         this.field_190665_b = p_190658_1_;
     }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftVex.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftVex.java
index 7f1776cb99ebb5e29a1b5e69ced5b99e40cf9ee0..6252fedc827a5dee33fb8f84789ce23eedcffe89 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftVex.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftVex.java
@@ -23,6 +23,10 @@ public class CraftVex extends CraftMonster implements Vex {
         MobEntity owner = getHandle().getOwner();
         return owner != null ? (Mob) owner.getBukkitEntity() : null;
     }
+
+    public void setSummoner(Mob summoner) {
+        getHandle().setOwner(summoner == null ? null : ((CraftMob) summoner).getHandle());
+    }
     // Paper end
 
     @Override
