From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: BillyGalbreath <Blake.Galbreath@GMail.com>
Date: Sun, 7 Oct 2018 00:54:21 -0500
Subject: [PATCH] Add sun related API


diff --git a/src/main/java/net/minecraft/entity/MobEntity.java b/src/main/java/net/minecraft/entity/MobEntity.java
index b47c6c9f75e2dbe705eb5ca25a272eafc67b5775..df36424317180110ec4885d88da8887950743935 100644
--- a/src/main/java/net/minecraft/entity/MobEntity.java
+++ b/src/main/java/net/minecraft/entity/MobEntity.java
@@ -1599,6 +1599,7 @@ public abstract class MobEntity extends LivingEntity {
 
     }
 
+    public boolean isInDaylight() { return this.func_204609_dp(); } // Paper - OBFHELPER
     protected boolean func_204609_dp() {
         if (this.field_70170_p.func_72935_r() && !this.field_70170_p.field_72995_K) {
             float f = this.func_70013_c();
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index f2545b295abb594efa5090a08a746f4b11e51e6c..d08f33887272a713e13e022051e392cfc15930ff 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -852,6 +852,13 @@ public class CraftWorld implements World {
         }
     }
 
+    // Paper start
+    @Override
+    public boolean isDayTime() {
+        return getHandle().func_72935_r();
+    }
+    // Paper end
+
     @Override
     public boolean createExplosion(double x, double y, double z, float power) {
         return createExplosion(x, y, z, power, false, true);
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftMob.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftMob.java
index a0261d39ce7fd4d3db8c7d6c1ef034f787b21880..d3d709cc6520510875a6b7f14ab429e5461b5683 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftMob.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftMob.java
@@ -78,4 +78,11 @@ public abstract class CraftMob extends CraftLivingEntity implements Mob {
     public long getSeed() {
         return getHandle().field_184653_bB;
     }
+
+    // Paper start
+    @Override
+    public boolean isInDaylight() {
+        return getHandle().isInDaylight();
+    }
+    // Paper end
 }
