From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: BillyGalbreath <Blake.Galbreath@GMail.com>
Date: Sun, 23 Sep 2018 20:59:53 -0500
Subject: [PATCH] Honor EntityAgeable.ageLock


diff --git a/src/main/java/net/minecraft/entity/AgeableEntity.java b/src/main/java/net/minecraft/entity/AgeableEntity.java
index 45a53b601655329907e0c71615da0ebc00fbfd62..e9793dd24fbf0088a48906a0b3176819a56e4e8f 100644
--- a/src/main/java/net/minecraft/entity/AgeableEntity.java
+++ b/src/main/java/net/minecraft/entity/AgeableEntity.java
@@ -81,6 +81,7 @@ public abstract class AgeableEntity extends CreatureEntity {
     }
 
     public void func_175501_a(int p_175501_1_, boolean p_175501_2_) {
+        if (ageLocked) return; // Paper - GH-1459
         int j = this.func_70874_b();
         int k = j;
 
