From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Gabriele C <sgdc3.mail@gmail.com>
Date: Mon, 22 Oct 2018 17:34:10 +0200
Subject: [PATCH] Add option to prevent players from moving into unloaded
 chunks #1551


diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index d8afb247f180b89f89cbed0f6e6c4fcc527799ff..6f410120018f95259e9e4f504ea1b588c2b4622b 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -390,4 +390,9 @@ public class PaperWorldConfig {
         waterOverLavaFlowSpeed = getInt("water-over-lava-flow-speed", 5);
         log("Water over lava flow speed: " + waterOverLavaFlowSpeed);
     }
+
+    public boolean preventMovingIntoUnloadedChunks = false;
+    private void preventMovingIntoUnloadedChunks() {
+        preventMovingIntoUnloadedChunks = getBoolean("prevent-moving-into-unloaded-chunks", false);
+    }
 }
diff --git a/src/main/java/net/minecraft/network/play/ServerPlayNetHandler.java b/src/main/java/net/minecraft/network/play/ServerPlayNetHandler.java
index b839fc791c310b3506dc73f9bcf21a1c60353b40..6765526357087f2c1ca09219e296aea3f946b353 100644
--- a/src/main/java/net/minecraft/network/play/ServerPlayNetHandler.java
+++ b/src/main/java/net/minecraft/network/play/ServerPlayNetHandler.java
@@ -538,6 +538,13 @@ public class ServerPlayNetHandler implements IServerPlayNetHandler {
                 }
                 speed *= 2f; // TODO: Get the speed of the vehicle instead of the player
 
+                // Paper start - Prevent moving into unloaded chunks
+                if (field_147369_b.field_70170_p.paperConfig.preventMovingIntoUnloadedChunks && worldserver.getChunkIfLoadedImmediately((int) Math.floor(p_184338_1_.func_187004_a()) >> 4, (int) Math.floor(p_184338_1_.func_187003_c()) >> 4) == null) {
+                    this.field_147371_a.func_179290_a(new SMoveVehiclePacket(entity));
+                    return;
+                }
+                // Paper end
+
                 if (d10 - d9 > Math.max(100.0D, Math.pow((double) (org.spigotmc.SpigotConfig.movedTooQuicklyMultiplier * (float) i * speed), 2)) && !this.func_217264_d()) {
                 // CraftBukkit end
                     ServerPlayNetHandler.field_147370_c.warn("{} (vehicle of {}) moved too quickly! {},{},{}", entity.func_200200_C_().getString(), this.field_147369_b.func_200200_C_().getString(), d6, d7, d8);
@@ -1136,9 +1143,9 @@ public class ServerPlayNetHandler implements IServerPlayNetHandler {
                         double d1 = this.field_147369_b.func_226278_cu_();
                         double d2 = this.field_147369_b.func_226281_cx_();
                         double d3 = this.field_147369_b.func_226278_cu_();
-                        double d4 = p_147347_1_.func_186997_a(this.field_147369_b.func_226277_ct_());
+                        double d4 = p_147347_1_.func_186997_a(this.field_147369_b.func_226277_ct_());double toX = d4; // Paper - OBFHELPER
                         double d5 = p_147347_1_.func_186996_b(this.field_147369_b.func_226278_cu_());
-                        double d6 = p_147347_1_.func_187000_c(this.field_147369_b.func_226281_cx_());
+                        double d6 = p_147347_1_.func_187000_c(this.field_147369_b.func_226281_cx_());double toZ = d6; // Paper - OBFHELPER
                         float f = p_147347_1_.func_186999_a(this.field_147369_b.field_70177_z);
                         float f1 = p_147347_1_.func_186998_b(this.field_147369_b.field_70125_A);
                         double d7 = d4 - this.field_184349_l;
@@ -1177,6 +1184,12 @@ public class ServerPlayNetHandler implements IServerPlayNetHandler {
                             } else {
                                 speed = field_147369_b.field_71075_bZ.field_75097_g * 10f;
                             }
+                            // Paper start - Prevent moving into unloaded chunks
+                            if (field_147369_b.field_70170_p.paperConfig.preventMovingIntoUnloadedChunks && (this.field_147369_b.func_226277_ct_() != toX || this.field_147369_b.func_226281_cx_() != toZ) && !worldserver.func_217354_b((int) Math.floor(toX) >> 4, (int) Math.floor(toZ) >> 4)) {
+                                this.internalTeleport(this.field_147369_b.func_226277_ct_(), this.field_147369_b.func_226278_cu_(), this.field_147369_b.func_226281_cx_(), this.field_147369_b.field_70177_z, this.field_147369_b.field_70125_A, Collections.emptySet());
+                                return;
+                            }
+                            // Paper end
 
                             if (!this.field_147369_b.func_184850_K() && (!this.field_147369_b.func_71121_q().func_82736_K().func_223586_b(GameRules.field_223615_r) || !this.field_147369_b.func_184613_cA())) {
                                 float f2 = this.field_147369_b.func_184613_cA() ? 300.0F : 100.0F;
