From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: BillyGalbreath <Blake.Galbreath@GMail.com>
Date: Sun, 7 Oct 2018 04:29:59 -0500
Subject: [PATCH] Add more Zombie API


diff --git a/src/main/java/net/minecraft/entity/monster/ZombieEntity.java b/src/main/java/net/minecraft/entity/monster/ZombieEntity.java
index 3418154c18852e6de1d755ed147e2cb22291ca16..47fde33848c571a0abd414ee38380677a103e91f 100644
--- a/src/main/java/net/minecraft/entity/monster/ZombieEntity.java
+++ b/src/main/java/net/minecraft/entity/monster/ZombieEntity.java
@@ -94,6 +94,7 @@ public class ZombieEntity extends MonsterEntity {
     private int field_204707_bD;
     public int field_204708_bE;
     private int lastTick = MinecraftServer.currentTick; // CraftBukkit - add field
+    private boolean shouldBurnInDay = true; // Paper
 
     public ZombieEntity(EntityType<? extends ZombieEntity> entitytypes, World world) {
         super(entitytypes, world);
@@ -262,6 +263,12 @@ public class ZombieEntity extends MonsterEntity {
         super.func_70636_d();
     }
 
+    // Paper start
+    public void stopDrowning() {
+        this.field_204708_bE = -1;
+        this.func_184212_Q().func_187227_b(ZombieEntity.field_204709_bA, false);
+    }
+    // Paper end
     public void func_204704_a(int p_204704_1_) {
         this.lastTick = MinecraftServer.currentTick; // CraftBukkit
         this.field_204708_bE = p_204704_1_;
@@ -287,9 +294,16 @@ public class ZombieEntity extends MonsterEntity {
 
     }
 
+    public boolean shouldBurnInDay() { return func_190730_o(); } // Paper - OBFHELPER
     protected boolean func_190730_o() {
-        return true;
+        return this.shouldBurnInDay; // Paper - use api value instead
+    }
+
+    // Paper start
+    public void setShouldBurnInDay(boolean shouldBurnInDay) {
+        this.shouldBurnInDay = shouldBurnInDay;
     }
+    // Paper end
 
     @Override
     public boolean func_70097_a(DamageSource p_70097_1_, float p_70097_2_) {
@@ -410,6 +424,7 @@ public class ZombieEntity extends MonsterEntity {
         p_213281_1_.func_74757_a("CanBreakDoors", this.func_146072_bX());
         p_213281_1_.func_74768_a("InWaterTime", this.func_70090_H() ? this.field_204707_bD : -1);
         p_213281_1_.func_74768_a("DrownedConversionTime", this.func_204706_dD() ? this.field_204708_bE : -1);
+        p_213281_1_.func_74757_a("Paper.ShouldBurnInDay", shouldBurnInDay); // Paper
     }
 
     @Override
@@ -421,7 +436,11 @@ public class ZombieEntity extends MonsterEntity {
         if (p_70037_1_.func_150297_b("DrownedConversionTime", 99) && p_70037_1_.func_74762_e("DrownedConversionTime") > -1) {
             this.func_204704_a(p_70037_1_.func_74762_e("DrownedConversionTime"));
         }
-
+        // Paper start
+        if (p_70037_1_.func_74764_b("Paper.ShouldBurnInDay")) {
+            shouldBurnInDay = p_70037_1_.func_74767_n("Paper.ShouldBurnInDay");
+        }
+        // Paper end
     }
 
     @Override
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftZombie.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftZombie.java
index 95fa706e711ce8485223da6957d86b9f46c80070..792c9fc50392259e7afb9990ea0b323f8ab5f0e4 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftZombie.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftZombie.java
@@ -94,6 +94,42 @@ public class CraftZombie extends CraftMonster implements Zombie {
     @Override
     public void setAgeLock(boolean b) {
     }
+    // Paper start
+    @Override
+    public boolean isDrowning() {
+        return getHandle().func_204706_dD();
+    }
+
+    @Override
+    public void startDrowning(int drownedConversionTime) {
+        getHandle().func_204704_a(drownedConversionTime);
+    }
+
+    @Override
+    public void stopDrowning() {
+        getHandle().stopDrowning();
+    }
+
+    @Override
+    public boolean shouldBurnInDay() {
+        return getHandle().shouldBurnInDay();
+    }
+
+    @Override
+    public boolean isArmsRaised() {
+        return getHandle().func_213398_dR();
+    }
+
+    @Override
+    public void setArmsRaised(final boolean raised) {
+        getHandle().func_213395_q(raised);
+    }
+
+    @Override
+    public void setShouldBurnInDay(boolean shouldBurnInDay) {
+        getHandle().setShouldBurnInDay(shouldBurnInDay);
+    }
+    // Paper end
 
     @Override
     public boolean getAgeLock() {
