From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: connorhartley <vectrixu+gh@gmail.com>
Date: Mon, 7 Jan 2019 14:43:48 -0600
Subject: [PATCH] Workaround for vehicle tracking issue on disconnect


diff --git a/src/main/java/net/minecraft/entity/player/ServerPlayerEntity.java b/src/main/java/net/minecraft/entity/player/ServerPlayerEntity.java
index 9127da97a66213703fbf25bc68e2e1b9fb1e4671..2133f5b4d42b95ac4687b4178daaec29ba1526b3 100644
--- a/src/main/java/net/minecraft/entity/player/ServerPlayerEntity.java
+++ b/src/main/java/net/minecraft/entity/player/ServerPlayerEntity.java
@@ -1502,6 +1502,13 @@ public class ServerPlayerEntity extends PlayerEntity implements IContainerListen
     public void func_71123_m() {
         this.field_193109_cv = true;
         this.func_184226_ay();
+
+        // Paper start - Workaround an issue where the vehicle doesn't track the passenger disconnection dismount.
+        if (this.func_184218_aH() && this.func_184187_bx() instanceof ServerPlayerEntity) {
+            this.func_184210_p();
+        }
+        // Paper end
+
         if (this.func_70608_bn()) {
             this.func_225652_a_(true, false);
         }
