From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <blake.galbreath@gmail.com>
Date: Wed, 9 Oct 2019 21:46:15 -0500
Subject: [PATCH] Add option to disable pillager patrols


diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index fde31f324b3578c65a7d064c466b80ee46b65e03..c51a149fd9cd7f2d250f07a4ca0e7ce8c9b91383 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -572,4 +572,9 @@ public class PaperWorldConfig {
     private void generatorSettings() {
         generateFlatBedrock = getBoolean("generator-settings.flat-bedrock", false);
     }
+
+    public boolean disablePillagerPatrols = false;
+    private void pillagerSettings() {
+        disablePillagerPatrols = getBoolean("game-mechanics.disable-pillager-patrols", disablePillagerPatrols);
+    }
 }
diff --git a/src/main/java/net/minecraft/world/spawner/PatrolSpawner.java b/src/main/java/net/minecraft/world/spawner/PatrolSpawner.java
index c700568573fb4b637fbbf5ff5d87777f534b43c3..f548487da853b7875967c3c9db1f59367ca734fe 100644
--- a/src/main/java/net/minecraft/world/spawner/PatrolSpawner.java
+++ b/src/main/java/net/minecraft/world/spawner/PatrolSpawner.java
@@ -24,6 +24,7 @@ public class PatrolSpawner implements ISpecialSpawner {
 
     @Override
     public int func_230253_a_(ServerWorld p_230253_1_, boolean p_230253_2_, boolean p_230253_3_) {
+        if (p_230253_1_.paperConfig.disablePillagerPatrols) return 0; // Paper
         if (!p_230253_2_) {
             return 0;
         } else if (!p_230253_1_.func_82736_K().func_223586_b(GameRules.field_230127_D_)) {
