From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Shane Freeder <theboyetronic@gmail.com>
Date: Tue, 24 Dec 2019 00:35:42 +0000
Subject: [PATCH] PlayerDeathEvent#shouldDropExperience


diff --git a/src/main/java/net/minecraft/entity/player/ServerPlayerEntity.java b/src/main/java/net/minecraft/entity/player/ServerPlayerEntity.java
index 1572bbb2de429468ad102ecf8280ccc2f582d244..9d0b4f028c2a3cf2f424e8863f377ba734169048 100644
--- a/src/main/java/net/minecraft/entity/player/ServerPlayerEntity.java
+++ b/src/main/java/net/minecraft/entity/player/ServerPlayerEntity.java
@@ -811,7 +811,7 @@ public class ServerPlayerEntity extends PlayerEntity implements IContainerListen
             this.func_241157_eT_();
         }
         // SPIGOT-5478 must be called manually now
-        this.func_226294_cV_();
+        if (event.shouldDropExperience()) this.func_226294_cV_(); // Paper - tie to event
         // we clean the player's inventory after the EntityDeathEvent is called so plugins can get the exact state of the inventory.
         if (!event.getKeepInventory()) {
             // Paper start - replace logic
