From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: nossr50 <nossr50@gmail.com>
Date: Thu, 26 Mar 2020 19:44:50 -0700
Subject: [PATCH] Add PlayerAttackEntityCooldownResetEvent


diff --git a/src/main/java/net/minecraft/entity/LivingEntity.java b/src/main/java/net/minecraft/entity/LivingEntity.java
index 1d84121ab4d2b0708c4916eea165324995c64483..44934e7eecb01fc33195fa0ed7bd850e73a2a9ac 100644
--- a/src/main/java/net/minecraft/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/entity/LivingEntity.java
@@ -1916,7 +1916,16 @@ public abstract class LivingEntity extends Entity {
 
             EntityDamageEvent event = CraftEventFactory.handleLivingEntityDamageEvent(this, damagesource, originalDamage, hardHatModifier, blockingModifier, armorModifier, resistanceModifier, magicModifier, absorptionModifier, hardHat, blocking, armor, resistance, magic, absorption);
             if (damagesource.func_76346_g() instanceof PlayerEntity) {
-                ((PlayerEntity) damagesource.func_76346_g()).func_184821_cY(); // Moved from EntityHuman in order to make the cooldown reset get called after the damage event is fired
+                // Paper start - PlayerAttackEntityCooldownResetEvent
+                if (damagesource.func_76346_g() instanceof ServerPlayerEntity) {
+                    ServerPlayerEntity player = (ServerPlayerEntity) damagesource.func_76346_g();
+                    if (new com.destroystokyo.paper.event.player.PlayerAttackEntityCooldownResetEvent(player.getBukkitEntity(), this.getBukkitEntity(), player.func_184825_o(0F)).callEvent()) {
+                        player.func_184821_cY();
+                    }
+                } else {
+                    ((PlayerEntity) damagesource.func_76346_g()).func_184821_cY();
+                }
+                // Paper end
             }
             if (event.isCancelled()) {
                 return false;
