From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sun, 7 Jun 2020 19:25:13 -0400
Subject: [PATCH] Use seed based lookup for Treasure Maps - Fixes lag from
 carto/sunken maps


diff --git a/src/main/java/net/minecraft/item/FilledMapItem.java b/src/main/java/net/minecraft/item/FilledMapItem.java
index 88d021b931bd5184b9df69138b96b9a3398a5e7b..1f106cfee606b0bc2a857a4068ea4de1b05d59f3 100644
--- a/src/main/java/net/minecraft/item/FilledMapItem.java
+++ b/src/main/java/net/minecraft/item/FilledMapItem.java
@@ -251,7 +251,7 @@ public class FilledMapItem extends AbstractMapItem {
 
                 for (l = 0; l < 128 * i; ++l) {
                     for (i1 = 0; i1 < 128 * i; ++i1) {
-                        abiomebase[l * 128 * i + i1] = p_226642_0_.func_226691_t_(new BlockPos((j / i - 64) * i + i1, 0, (k / i - 64) * i + l));
+                        abiomebase[l * 128 * i + i1] = p_226642_0_.getBiomeBySeed((j / i - 64) * i + i1, 0, (k / i - 64) * i + l); // Paper
                     }
                 }
 
diff --git a/src/main/java/net/minecraft/world/server/ServerWorld.java b/src/main/java/net/minecraft/world/server/ServerWorld.java
index 427974cd5658d88b4d2bbf85571c05ccfb123b56..319fa6f361b8231617102b5808f2d180e67f5ba6 100644
--- a/src/main/java/net/minecraft/world/server/ServerWorld.java
+++ b/src/main/java/net/minecraft/world/server/ServerWorld.java
@@ -414,8 +414,8 @@ public class ServerWorld extends World implements ISeedReader {
         this.field_241103_E_.func_76069_a(p_241113_4_);
     }
 
-    @Override
-    public Biome func_225604_a_(int p_225604_1_, int p_225604_2_, int p_225604_3_) {
+    public Biome getBiomeBySeed(int i, int j, int k) { return func_225604_a_(i, j, k); } // Paper - OBFHELPER
+    @Override public Biome func_225604_a_(int p_225604_1_, int p_225604_2_, int p_225604_3_) {
         return this.func_72863_F().func_201711_g().func_202090_b().func_225526_b_(p_225604_1_, p_225604_2_, p_225604_3_);
     }
 
