From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mariell Hoversholm <proximyst@proximyst.com>
Date: Sun, 23 Aug 2020 19:01:04 +0200
Subject: [PATCH] Don't require FACING data


diff --git a/src/main/java/net/minecraft/dispenser/DefaultDispenseItemBehavior.java b/src/main/java/net/minecraft/dispenser/DefaultDispenseItemBehavior.java
index f04714ae5e1cb63da9598cc032f47144571c6cad..bf22046a9a46e9c984864984e97a30067057e3fc 100644
--- a/src/main/java/net/minecraft/dispenser/DefaultDispenseItemBehavior.java
+++ b/src/main/java/net/minecraft/dispenser/DefaultDispenseItemBehavior.java
@@ -12,20 +12,22 @@ import org.bukkit.event.block.BlockDispenseEvent;
 // CraftBukkit end
 
 public class DefaultDispenseItemBehavior implements IDispenseItemBehavior {
+    private Direction enumdirection; // Paper
 
     public DefaultDispenseItemBehavior() {}
 
     @Override
     public final ItemStack dispense(IBlockSource isourceblock, ItemStack itemstack) {
+        enumdirection = isourceblock.func_189992_e().func_177229_b(DispenserBlock.field_176441_a); // Paper - cache facing direction
         ItemStack itemstack1 = this.func_82487_b(isourceblock, itemstack);
 
         this.func_82485_a(isourceblock);
-        this.func_82489_a(isourceblock, (Direction) isourceblock.func_189992_e().func_177229_b(DispenserBlock.field_176441_a));
+        this.func_82489_a(isourceblock, enumdirection); // Paper - cache facing direction
         return itemstack1;
     }
 
     protected ItemStack func_82487_b(IBlockSource p_82487_1_, ItemStack p_82487_2_) {
-        Direction enumdirection = (Direction) p_82487_1_.func_189992_e().func_177229_b(DispenserBlock.field_176441_a);
+        // Paper - cached enum direction
         IPosition iposition = DispenserBlock.func_149939_a(p_82487_1_);
         ItemStack itemstack1 = p_82487_2_.func_77979_a(1);
 
