From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <Blake.Galbreath@GMail.com>
Date: Sat, 3 Oct 2020 22:00:27 -0500
Subject: [PATCH] Fix deop kicking non-whitelisted player when white list is
 not enabled


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index d2e266754541810c5e33ba0427505fe18d22e3e7..2c7f3f6406b3e5f73dfd92303b77495a9532776d 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -2005,6 +2005,7 @@ public abstract class MinecraftServer extends RecursiveEventLoop<TickDelayedTask
         if (this.func_205744_aT()) {
             PlayerList playerlist = p_205743_1_.func_197028_i().func_184103_al();
             WhiteList whitelist = playerlist.func_152599_k();
+            if (!((DedicatedServer)getServer()).func_213221_d_().field_219006_S.get()) return; // Paper - white list not enabled
             List<ServerPlayerEntity> list = Lists.newArrayList(playerlist.func_181057_v());
             Iterator iterator = list.iterator();
 
