From a702e991854bea90d1d75de133f55f8748715f2c Mon Sep 17 00:00:00 2001
From: heroslender <bruno.martins1998@hotmail.com>
Date: Fri, 4 Dec 2020 20:31:15 +0000
Subject: [PATCH] Fix player joining with null address


diff --git a/src/main/java/net/minecraft/server/PlayerList.java b/src/main/java/net/minecraft/server/PlayerList.java
index 3c2d3197f..e943b9a7e 100644
--- a/src/main/java/net/minecraft/server/PlayerList.java
+++ b/src/main/java/net/minecraft/server/PlayerList.java
@@ -36,6 +36,7 @@ import org.bukkit.event.player.PlayerQuitEvent;
 import org.bukkit.event.player.PlayerRespawnEvent;
 import org.bukkit.event.player.PlayerTeleportEvent.TeleportCause;
 import org.bukkit.util.Vector;
+import org.github.paperspigot.PaperSpigotConfig;
 import org.spigotmc.event.player.PlayerSpawnLocationEvent;
 // CraftBukkit end
 
@@ -433,6 +434,12 @@ public abstract class PlayerList {
         // in the event, check with plugins to see if it's ok, and THEN kick
         // depending on the outcome.
         SocketAddress socketaddress = loginlistener.networkManager.getSocketAddress();
+        // Paper start - Fix null address(#1)
+        if (PaperSpigotConfig.exploitNullAddressEnabled && ((java.net.InetSocketAddress) socketaddress).getAddress() == null) {
+            loginlistener.disconnect(PaperSpigotConfig.exploitNullAddressKickMessage);
+            return null;
+        }
+        // Paper end
 
         EntityPlayer entity = new EntityPlayer(server, server.getWorldServer(0), gameprofile, new PlayerInteractManager(server.getWorldServer(0)));
         Player player = entity.getBukkitEntity();
diff --git a/src/main/java/org/github/paperspigot/PaperSpigotConfig.java b/src/main/java/org/github/paperspigot/PaperSpigotConfig.java
index d6d9899e2..65f43b002 100644
--- a/src/main/java/org/github/paperspigot/PaperSpigotConfig.java
+++ b/src/main/java/org/github/paperspigot/PaperSpigotConfig.java
@@ -222,4 +222,11 @@ public class PaperSpigotConfig
     {
         warnForExcessiveVelocity = getBoolean("warnWhenSettingExcessiveVelocity", true);
     }
+
+    public static boolean exploitNullAddressEnabled;
+    public static String exploitNullAddressKickMessage;
+    private static void exploits() {
+        exploitNullAddressEnabled = getBoolean("exploits.null-address.enabled", true);
+        exploitNullAddressKickMessage = getString("exploits.null-address.kick-message", "Invalid address!");
+    }
 }
-- 
2.28.0

